### DISCUSSION TITLE
NC mesh with hanging element?

### DISCUSSION THREAD

**Discussion by dongzhangim** [mfem/mfem#4632] [Q&A] [answered]
hello, all
i'm new to mfem. I wanna use MFEM for the simulation of elastic deformation of a nonconforming mesh with hanging element. Hanging element means when a quad or hex element is refined, some subdivided elements are deleted leaving the remaining ones(one) hanging. That's my understanding.

First, I tested two elements with one refined (see Figure A). The results (generated by ex21) are shown in Figure B.
![图片1](https://github.com/user-attachments/assets/4b99780b-140d-4c21-bc47-7ce1a8c51233)
The mesh files are as follows.

```
MFEM NC mesh v1.0

dimension
2

elements
6
0 1 3 0 0 1 4 3
0 2 3 3 2 3 4 5
0 2 3 0 1 8 10 7
0 2 3 0 8 2 9 10
0 2 3 0 7 10 6 4
0 2 3 0 10 9 5 6

boundary
9
3 1 0 1
3 1 1 8
3 1 2 8
3 1 3 4
3 1 4 6
3 1 5 6
1 1 0 3
2 1 2 8
2 1 5 9

vertex_parents
5
6 4 5
7 1 4
8 1 2
9 2 5
10 6 8

coordinates
6
2
0 0
1 0
2 0
0 1
1 1
2 1

mfem_mesh_end
```

Second, I try to delete the element with id 5 by removing its id from the element of id 1 as depicted in Figure C, however, it does not work with errors:

_Verification failed: (geom == Geometry::SEGMENT || geom == Geometry::TRIANGLE || geom == Geometry::SQUARE || geom == Geometry::CUBE || geom == Geometry::PRISM || geom == Geometry::PYRAMID || geom == Geometry::TETRAHEDRON) is false:
 --&gt; Element type 0 is not supported by NCMesh.
 ... in function: void mfem::CheckSupportedGeom(Geometry::Type)
 ... in file: mesh/ncmesh.cpp:128_

The NC mesh files are as follows:

```
MFEM NC mesh v1.0
dimension
2

elements
5
0 1 3 0 0 1 4 3
0 2 3 3 2 3 4
0 2 3 0 1 8 10 7
0 2 3 0 8 2 9 10
0 2 3 0 7 10 6 4

boundary
9
3 1 0 1
3 1 1 8
3 1 2 8
3 1 3 4
3 1 4 6
3 1 10 6
1 1 0 3
2 1 2 8
2 1 10 9

vertex_parents
5
6 4 5
7 1 4
8 1 2
9 2 5
10 6 8

coordinates
6
2
0 0
1 0
2 0
0 1
1 1
2 1

mfem_mesh_end
```

So, how can we tackle nc mesh with hanging elements?

**Comment by v-dobrev** [comment, marked as answer]
One problem that I see is: vertex id 5 is no longer needed, so you can remove it from `coordinates` and also from `vertex_parents`. Also, some of the vertices that you previously described in the `vertex_parents` section have to be added as regular vertices due to the removal of vertex with id 5.

**Reply by dongzhangim** [reply]
Dear Veselin,
Thanks for your help. After deleting the hanging node with an ID of 5 and making two nodes with coordinates of (1.5, 1) and (2, 0.5) regular, it works as shown below, although the refinement hierarchy cannot be stored. 
![图片1](https://github.com/user-attachments/assets/2caee519-2a28-4fd4-a131-9651266cba77)
```
MFEM NC mesh v1.0
dimension
2

elements
4
0 1 3 0 0 1 4 3
0 2 3 0 1 8 9 7
0 2 3 0 8 2 6 9
0 2 3 0 7 9 5 4

boundary
9
3 1 0 1
3 1 1 8
3 1 2 8
3 1 3 4
3 1 4 5
3 1 5 9
3 1 6 9
1 1 0 3
2 1 2 6

vertex_parents
3
7 1 4
8 1 2
9 5 8

coordinates
7
2
0 0
1 0
2 0
0 1
1 1
1.5 1
2 0.5

mfem_mesh_end

```

Also, we can make the single node with the coordinate of (1 0.5) nonregular, it works as well. 

```
MFEM NC mesh v1.0
dimension
2

elements
4
0 1 3 0 0 1 5 4
0 2 3 0 1 2 7 9
0 2 3 0 2 3 8 7
0 2 3 0 9 7 6 5

boundary
9
3 1 0 1
3 1 1 2
3 1 2 3
3 1 4 5
3 1 5 6
3 1 6 7
3 1 7 8
1 1 0 4
2 1 3 8

vertex_parents
1
9 1 5

coordinates
9
2
0 0
1 0
1.5 0
2 0
0 1
1 1
1.5 1
1.5 0.5
2 0.5

mfem_mesh_end
```

**Comment by dongzhangim** [comment]
is it possible to add a hanging point onto another hanging point?  as follows.   // m is a hanging node of p1 and p2    // q is a hanging node of p1 and m 



<img width="449" height="401" alt="图片1" src="https://github.com/user-attachments/assets/e0ea2fd8-90c3-47a5-8ffd-bc89b22ef65c" />

I've tried the following codes, it gives error when running NCMesh::NCMesh(const Mesh *mesh)

```
#include "mfem.hpp"
#include <fstream>

using namespace mfem;

int main()
{
   Mesh mesh(2, 0, 0);

   int p0 = mesh.AddVertex(0, 0);
   int p1 = mesh.AddVertex(1, 0);
   int p2 = mesh.AddVertex(1, 1);
   int p3 = mesh.AddVertex(0, 1);

   int m  = mesh.AddVertex(1, 0.5);
   int q = mesh.AddVertex(1, 0.25);

   int r0 = mesh.AddVertex(1.25, 0);
   int r1 = mesh.AddVertex(1.25, 0.25);


   // the left is a 1*1 element
   mesh.AddQuad(p0, p1, p2, p3);

   //  the right is a 1/4*1/4 element
   mesh.AddQuad(p1, r0, r1, q);

   // m is a hanging node of p1 and p2 
   mesh.AddVertexParents(m,  p1, p2); // 1/2
   // q is a hanging node of p1 and m 
   mesh.AddVertexParents(q, p1, m);  // 1/4

   mesh.FinalizeMesh(0);              // 0：no refinement
   mesh.Save("quarter.mesh");

   return 0;
}
```

**Reply by v-dobrev** [reply]
If the goal is to have the point `q` to be 1/4 the way from `p1` to `p2`, then there's not need to introduce `m` -- the feature that allows for that was added in #4371. @dylan-copeland can provide more details about constructing such a mesh in your code.

**Reply by dylan-copeland** [reply]
For nonuniform spacing in an `NCMesh`, see `miniapps/meshing/ref321.cpp` introduced by PR 4371, e.g. `Refinement(elem, type, 0.25)` for 1/4 spacing. The easiest way to generate an `NCMesh` like this is probably to start with a coarse conforming mesh and add refinements of elements as in `ref321.cpp`.

**Reply by v-dobrev** [reply]
@dylan-copeland, what if the whole mesh is just the two elements from the illustration? Do we have a method `AddVertexParents` where we can specify the 1/4 spacing directly, modifying the sample code above?

**Reply by dylan-copeland** [reply]
We have protected methods in `NCMesh` to set spacing, but the public interface only supports refinement as in the miniapp `ref321`. Element refinement is the intended way to produce an `NCMesh`.

**Reply by dongzhangim** [reply]
what we want is to generate a NCmesh from an octree or b+tree voxels. Therefore, the AddVertexParents function which can specify the 1/4 spacing directly will be more interesting. Otherwise, i have to conduct MFEM refinement during the bool operation for the octree or b+ tree. Also during the refinement, some voxel may be deleted. I'm not sure whether the some child element can be deleted during the refinement.

**Reply by dylan-copeland** [reply]
It may be possible to do this with refinement and derefinement, although not optimal. If new capabilities would be more convenient, you can make a PR with the necessary code changes.
